File: prettier-plugin-csharpier/src/Printer/Types/ElseClause.ts
import { Doc } from "prettier";
import { PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, printSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { BlockNode } from "./Block";
import { IfStatementNode } from "./IfStatement";

export interface ElseClauseNode extends SyntaxTreeNode<"ElseClause"> {
    elseKeyword?: SyntaxToken;
    statement?: SyntaxTreeNode;
}

export const printElseClause: PrintMethod<ElseClauseNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [printPathSyntaxToken(path, "elseKeyword")];
    let statement = path.call(print, "statement");
    if (node.statement?.nodeType === "Block") {
        parts.push(statement);
    } else if (node.statement?.nodeType === "IfStatement") {
        parts.push(" ", statement);
    } else {
        parts.push(indent(concat([hardline, statement])));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/ElseClause.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class ElseClause
{
    public static Doc Print(ElseClauseSyntax node, PrintingContext context)
    {
        return Doc.Concat(
            Token.Print(node.ElseKeyword, context),
            node.Statement is IfStatementSyntax ifStatementSyntax
                ? Doc.Concat(" ", IfStatement.Print(ifStatementSyntax, context))
                : OptionalBraces.Print(node.Statement, context)
        );
    }
}
